### ISSUE TITLE
Subscriptions aren't re-submitted properly on reconnection after connection failures

### ISSUE THREAD

**Issue by Shahor** [enisdenjo/graphql-ws#85]
**Screenshot**
First websocket and its related subscriptions

![image](https://user-images.githubusercontent.com/204561/101792231-c2db6c00-3b04-11eb-9c0d-0903ba71d3b9.png)

Next valid connection with its related subscriptions:

![image](https://user-images.githubusercontent.com/204561/101792280-cf5fc480-3b04-11eb-84a5-96e1addce7c3.png)

As you can see, the second screenshot has much less subscriptions sent to the server.

**Expected Behaviour**

No subscriptions are lost while reconnecting to a server

**Actual Behaviour**

It managed to reconnect to the server properly but didn't send the whole set of subscriptions

**Debug Information**

This seems to happen *only* when the websockets aren't able to reconnect right away to a server.
In my case: I restarted my server locally and because it's using typescript and transpiling on the fly with babel, it takes a little while to restart properly. So to reproduce I just launched my client, my server and then restarted my server and waited.

I am joining a `har` file below that has content as well, you can open this in any Chromium browser (sadly Firefox doesn't handle HAR with content yet) by dragging the file into your `dev tools network` panel.

It should load the whole story and you would be able to see my network at the moment of the bug happening.

[websockets-operations-after-reconnection.har.zip](https://github.com/enisdenjo/graphql-ws/files/5673806/websockets-operations-after-reconnection.har.zip)

**Further Information**

As with the previous issue, please let me know if you need more information and/or help 🙏 

Thank you again for your great work on this new library!

**Comment by enisdenjo** [comment]
Hey hey, thanks for reporting this and for the detailed explanation! This looks serious...

I'll start of by asking: are you sure that all subscribers are still active/streaming/mounted when failing? [All subscribers are basically stuck in an infinite "subscribe" loop until one of the following happens](https://github.com/enisdenjo/graphql-ws/blob/bf1b5373521f91faccc52f9eb3a80705e8f32a8e/src/client.ts#L589-L626):
- you complete the sink/observer (using the cleanup function from the `const cleanup = client.subscribe(...)`)
- you gracefully shut the client down (using `client.dispose()`)
- retry attempts are surpassed, error is reported to all subscribers

**Comment by enisdenjo** [comment]
I think I figured it out! PR incoming...

**Comment by enisdenjo** [comment]
Hmm, it was not what I thought it was...

Furthermore, I added a test that [should resubscribe all subscribers on silent reconnects](https://github.com/enisdenjo/graphql-ws/blob/e527452c38d8d2724f0af581f75b6b99c98f6e96/src/tests/client.ts#L702-L749). It is passing.

Can you elaborate a bit more? A repro would be perfect!

**Comment by Shahor** [comment]
> I'll start of by asking: are you sure that all subscribers are still active/streaming/mounted when failing?

I will try to make sure of that tomorrow, I can't say for sure right now.

What I can tell for sure is if the socket is able to reconnect right away then all subscriptions are re-submitted without issues.

This error only happens whenever the socket cannot reconnect right away.

**Comment by enisdenjo** [comment]
I released [v3.0.2](https://github.com/enisdenjo/graphql-ws/releases/tag/v3.0.2) which includes a one small retrying fix; still, if you could update before continuing the bug hunt tomorrow, that would be great!

> This error only happens whenever the socket cannot reconnect right away.

Ok, that is helpful for now - I'll try brining up a failing test in the meantime.

**Comment by Shahor** [comment]
> I released [v3.0.2](https://github.com/enisdenjo/graphql-ws/releases/tag/v3.0.2) which includes a one small retrying fix; still, if you could update before continuing the bug hunt tomorrow, that would be great!
> 
> > This error only happens whenever the socket cannot reconnect right away.
> 
> Ok, that is helpful for now - I'll try bringing up a failing test in the meantime.

Of course, I will retry with that version tomorrow no problem :)
Thanks for your help!

**Comment by Shahor** [comment]
Just tried with the updated version, it still seems to be happening when reconnection doesn't happen right away.

[websockets-operations-after-reconnection-2.har.zip](https://github.com/enisdenjo/graphql-ws/files/5678374/websockets-operations-after-reconnection-2.har.zip)

**Comment by enisdenjo** [comment]
I've managed to find some revelations in the meantime, bear with me - a fix is coming!

**Comment by enisdenjo** [comment]
Hey @Shahor, I have released a new version which, hopefully, fixes this issue.

[I've also added a more comprehensive, retry with timeout delay, test for ~50 active subscribers](https://github.com/enisdenjo/graphql-ws/blob/2aa9c41b79556c061f2354c4157725d6c39b9c67/src/tests/client.ts#L833-L880) - it is passing too.

If it turns out that `v3.1.0` does not fix the issue at hand, I'd kindly ask you to provide a repro or, maybe, a failing test. The WebSocket logs are not very useful for debugging this - I need info about currently active subscribers, reported errors/completes, etc.

**Comment by Shahor** [comment]
Sadly this doesn't fix the issue :( 
I will try to provide a reproducible repo so we can move forward with this. 

Thank you very much for your help again

**Comment by enisdenjo** [comment]
Damn, sad to hear that! Reopening the issue then.

Please share the repro once you have it, I'll investigate as soon as I can!

**Comment by enisdenjo** [comment]
@Shahor any news on this? Have you managed to write up a repro?

**Comment by Shahor** [comment]
> @Shahor any news on this? Have you managed to write up a repro?

Hey @enisdenjo 👋 

Sorry, I was on time off, I'll get back to this asap with a repro (hopefully).

**Comment by enisdenjo** [comment]
No probs! Hope you enjoyed your rest. 😄 

Looking forward to more info!

**Comment by enisdenjo** [comment]
As no one else reported experiencing this behaviour, I am closing the issue due to inactivity. @Shahor, if you manage to get a repro - please reopen it, I'll be more than happy to come back and re-evaluate.

**Comment by enisdenjo** [comment]
Hey @Shahor, I've done a rewrite of the client internals in #100. Would be great if you could try out the [`v4.1.2` release](https://github.com/enisdenjo/graphql-ws/releases/tag/v4.1.2) and check if this issue still presents itself.

**Comment by Shahor** [comment]
Hey @enisdenjo 

Thanks for the update! I will take some time today to check that.
Thank you so much for getting back to me on this, I am very sorry I didn't have any moment to try and reproduce on a smaller repo.
I haven't forgotten though and will get back on it as soon as possible.

In the meantime I'll try your update and will let you know!

**Comment by Shahor** [comment]
I come bearing good news 😅 

<img width="925" alt="image" src="https://user-images.githubusercontent.com/204561/105703593-c3da2880-5f0d-11eb-9a0d-aeff2abd0f86.png">

As you can see on the left it took some time for my server to restart (intended) and it seems all subscriptions were re-established properly. 

So I'd say that rewrite brought some good stuff along!

Great stuff man, thanks so much for your help! 👏 
I'll keep an eye on this and will let you know in case I see something weird.

**Comment by enisdenjo** [comment]
Wohoo! 🍾 

Very nice news, thanks for sharing. 😄

**Comment by Shahor** [comment]
Hey @enisdenjo 👋🏽 

I'm very sorry to be the bearer of bad news, but it seems this is coming back :( 

We're using graphql-ws 5.5.5 at the moment. 
Here's an attached `har` file you can open in chrome (firefox won't show content) and see the messages exchanges. 

[slite.shahor.slite.dev.har.zip](https://github.com/enisdenjo/graphql-ws/files/10246386/slite.shahor.slite.dev.har.zip)
_Had to zip or github would refuse the upload_

This happens when our server goes away and we reconnect after a few seconds.

**Comment by enisdenjo** [comment]
@Shahor hey there, does the same issue happen with the latest version of the client? The `har` file is kinda hard to follow, would it be too much to ask for a repro? Thanks in advance!
